### PR TITLE
Fix make_flux_sparsity_pattern() to use internal and flux masks in parallel

### PR DESCRIPTION
**Pull request by spco**
Relies on https://github.com/dealii/dealii/pull/2846 to reduce the memory consumption for dG quite dramatically when in parallel. I'm submitting this now to check that it at least builds (it does on my system), if someone would oblige?

This doesn't really merit a WIP tag as it's (hopefully) working, but no point merging before https://github.com/dealii/dealii/pull/2846 is merged.

### CODE REVIEW

**Comment by spco** [comment]
Addresses #1112

**Comment by spco** [comment]
On my machine, 4 MPI processes, with DG temperature, FEM compositions, output at timestep 100 on a modified composition-passive test gives
Old version:
```
Total system matrix memory consumption: 18.25 MB.
Total system matrix nnz: 1574664
system matrix nnz by block: 
      176698       30987      139257       90871       90871
       30987        5253       26163       16030       16030
      139257       26163       97605       71739       71739
       90871       16030       71739       46938       46929
       90871       16030       71739       46929       46938

Total system preconditioner matrix memory consumption: 18.25 MB.
Total system preconditioner matrix nnz: 1574664
system preconditioner matrix nnz by block: 
      176698       30987      139257       90871       90871
       30987        5253       26163       16030       16030
      139257       26163       97605       71739       71739
       90871       16030       71739       46938       46929
       90871       16030       71739       46929       46938


Number of active cells: 250 (on 5 levels)
Number of degrees of freedom: 6818 (2142+284+2250+1071+1071)



+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |      91.5s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble Stokes system          |       102 |      16.7s |        18% |
| Assemble composition system     |       204 |      22.6s |        25% |
| Assemble temperature system     |       102 |      20.5s |        22% |
| Build Stokes preconditioner     |       102 |      12.4s |        14% |
| Build composition preconditioner|       204 |     0.493s |      0.54% |
| Build temperature preconditioner|       102 |     0.428s |      0.47% |
| Solve Stokes system             |       102 |      6.14s |       6.7% |
| Solve composition system        |       204 |     0.723s |      0.79% |
| Solve temperature system        |       102 |     0.338s |      0.37% |
| Initialization                  |         1 |     0.128s |      0.14% |
| Postprocessing                  |       101 |      1.04s |       1.1% |
| Refine mesh structure, part 1   |        21 |      2.31s |       2.5% |
| Refine mesh structure, part 2   |        21 |      0.12s |      0.13% |
| Setup dof systems               |        22 |      6.28s |       6.9% |
| Setup initial conditions        |         2 |    0.0637s |      0.07% |
+---------------------------------+-----------+------------+------------+
```
New version:
```
Timestep 100
Total system matrix memory consumption: 2.71 MB.
Total system matrix nnz: 216930
system matrix nnz by block: 
       61586       12450           0           0           0
       12450           3           0           0           0
           0           0       97605           0           0
           0           0           0       16418           0
           0           0           0           0       16418

Total system preconditioner matrix memory consumption: 2.11 MB.
Total system preconditioner matrix nnz: 163828
system preconditioner matrix nnz by block: 
       31022           0           0           0           0
           0        2365           0           0           0
           0           0       97605           0           0
           0           0           0       16418           0
           0           0           0           0       16418


Number of active cells: 250 (on 5 levels)
Number of degrees of freedom: 6818 (2142+284+2250+1071+1071)



+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |      76.5s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble Stokes system          |       102 |      14.3s |        19% |
| Assemble composition system     |       204 |      19.9s |        26% |
| Assemble temperature system     |       102 |      18.4s |        24% |
| Build Stokes preconditioner     |       102 |       9.9s |        13% |
| Build composition preconditioner|       204 |     0.214s |      0.28% |
| Build temperature preconditioner|       102 |     0.395s |      0.52% |
| Solve Stokes system             |       102 |      3.37s |       4.4% |
| Solve composition system        |       204 |     0.711s |      0.93% |
| Solve temperature system        |       102 |     0.333s |      0.43% |
| Initialization                  |         1 |     0.146s |      0.19% |
| Postprocessing                  |       101 |     0.995s |       1.3% |
| Refine mesh structure, part 1   |        21 |      2.08s |       2.7% |
| Refine mesh structure, part 2   |        21 |     0.118s |      0.15% |
| Setup dof systems               |        22 |      4.36s |       5.7% |
| Setup initial conditions        |         2 |    0.0721s |     0.094% |
+---------------------------------+-----------+------------+------------+
```

So a not insignificant computational saving on a small testcase, across both assemblies and solves. I'm sure a larger testcase would bear this out.

Of course the big win is in terms of memory usage.

**Comment by yinghe616** [comment]
I think your test case is just too easy. The solve_stokes doesn't cost too
much of the total time. If you run on a harder problem, saying solve stokes
takes 20%-30% total time, then you will see the big difference :)

**Comment by spco** [comment]
Sure, I would certainly expect that, but even on this easy testcase we see a 15% wallclock improvement, so it will only get better on larger problems!

**Comment by yinghe616** [comment]
Yes, I'd like to use this parallel version once it's merged.:)
BTW,  could you explain why the pressure-pressure block has the non-zero coupling? is it due to some constraint?

**Comment by spco** [comment]
Great! Obviously it requires the deal.II development version until 8.5 is released.

A run with shell_simple_2d.prm, with dG temperature, gives at timestep 50:
Old: 
```
Number of active cells: 1416 (on 8 levels)
Number of degrees of freedom: 26841 (12490+1607+12744)
```
```
*** Timestep 50:  t=3.83212e+06 years
   Solving temperature system... 9 iterations.
   Solving Stokes system... 30+2 iterations.

   Postprocessing:
     Writing graphical output:           output/solution-00050
     RMS, max velocity:                  0.448 m/year, 0.887 m/year
     Temperature min/avg/max:            975.4 K, 2554 K, 4860 K
     Heat fluxes through boundary parts: 2.301e+05 W, 1.34e+05 W, 1280 W, 1280 W
     System matrix memory consumption:   11.08 MB
                                         
Total system matrix memory consumption: 43.69 MB.
Total system matrix nnz: 3787943
system matrix nnz by block: 
      969788      166551      793530
      166551       27801      147546
      793530      147546      575100

Total system preconditioner matrix memory consumption: 43.69 MB.
Total system preconditioner matrix nnz: 3787943
system preconditioner matrix nnz by block: 
      969788      166551      793530
      166551       27801      147546
      793530      147546      575100

+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |       187s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble Stokes system          |        54 |      15.5s |       8.3% |
| Assemble temperature system     |        54 |      91.6s |        49% |
| Build Stokes preconditioner     |         7 |      3.66s |         2% |
| Build temperature preconditioner|        54 |      2.11s |       1.1% |
| Solve Stokes system             |        54 |      30.3s |        16% |
| Solve temperature system        |        54 |      1.45s |      0.77% |
| Initialization                  |         1 |     0.104s |     0.056% |
| Postprocessing                  |        51 |      24.5s |        13% |
| Refine mesh structure, part 1   |         6 |      2.11s |       1.1% |
| Refine mesh structure, part 2   |         6 |     0.108s |     0.058% |
| Setup dof systems               |         7 |      6.36s |       3.4% |
| Setup initial conditions        |         4 |      0.45s |      0.24% |
+---------------------------------+-----------+------------+------------+
```
New:
```
Number of active cells: 1416 (on 8 levels)
Number of degrees of freedom: 26841 (12490+1607+12744)
```
```
*** Timestep 50:  t=3.83212e+06 years
   Solving temperature system... 9 iterations.
   Solving Stokes system... 30+2 iterations.

   Postprocessing:
     Writing graphical output:           output/solution-00050
     RMS, max velocity:                  0.448 m/year, 0.887 m/year
     Temperature min/avg/max:            975.4 K, 2554 K, 4860 K
     Heat fluxes through boundary parts: 2.301e+05 W, 1.34e+05 W, 1280 W, 1280 W
     System matrix memory consumption:   3.10 MB
                                         
Total system matrix memory consumption: 12.27 MB.
Total system matrix nnz: 1042280
system matrix nnz by block: 
      334404       66288           0
       66288         200           0
           0           0      575100

Total system preconditioner matrix memory consumption: 9.01 MB.
Total system preconditioner matrix nnz: 757513
system preconditioner matrix nnz by block: 
      169898           0           0
           0       12515           0
           0           0      575100

+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |       147s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble Stokes system          |        54 |      13.2s |       8.9% |
| Assemble temperature system     |        54 |        81s |        55% |
| Build Stokes preconditioner     |         7 |      2.69s |       1.8% |
| Build temperature preconditioner|        54 |       1.7s |       1.2% |
| Solve Stokes system             |        54 |      9.83s |       6.7% |
| Solve temperature system        |        54 |      1.18s |       0.8% |
| Initialization                  |         1 |     0.217s |      0.15% |
| Postprocessing                  |        51 |      22.5s |        15% |
| Refine mesh structure, part 1   |         6 |      1.83s |       1.2% |
| Refine mesh structure, part 2   |         6 |     0.085s |     0.058% |
| Setup dof systems               |         7 |      4.84s |       3.3% |
| Setup initial conditions        |         4 |     0.338s |      0.23% |
+---------------------------------+-----------+------------+------------+
```
so we really begin to see some improvement in the Stokes solve, but also improvement in the assemblies and even other things like setting up the dof systems.

What's actually quite interesting is to compare to the FEM. Here's the output at 50 timesteps again:
```
Number of active cells: 3270 (on 8 levels)
Number of degrees of freedom: 46706 (28694+3665+14347)
```
```
*** Timestep 50:  t=3.44344e+06 years
   Solving temperature system... 12 iterations.
   Solving Stokes system... 27+0 iterations.

   Postprocessing:
     Writing graphical output:           output/solution-00050
     RMS, max velocity:                  0.402 m/year, 0.799 m/year
     Temperature min/avg/max:            973 K, 2532 K, 4466 K
     Heat fluxes through boundary parts: 5.918e+04 W, 2.125e+05 W, 1702 W, 1702 W
     System matrix memory consumption:   3.97 MB
                                         
Total system matrix memory consumption: 15.49 MB.
Total system matrix nnz: 1303317
system matrix nnz by block: 
      787372      154624           0
      154624         478           0
           0           0      206219

Total system preconditioner matrix memory consumption: 7.86 MB.
Total system preconditioner matrix nnz: 637098
system preconditioner matrix nnz by block: 
      402178           0           0
           0       28701           0
           0           0      206219

+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |       144s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble Stokes system          |        54 |      15.6s |        11% |
| Assemble temperature system     |        54 |      70.2s |        49% |
| Build Stokes preconditioner     |         7 |      3.53s |       2.4% |
| Build temperature preconditioner|        54 |     0.395s |      0.27% |
| Solve Stokes system             |        54 |      10.3s |       7.2% |
| Solve temperature system        |        54 |     0.696s |      0.48% |
| Initialization                  |         1 |     0.125s |     0.087% |
| Postprocessing                  |        51 |      27.2s |        19% |
| Refine mesh structure, part 1   |         6 |      2.19s |       1.5% |
| Refine mesh structure, part 2   |         6 |     0.125s |     0.086% |
| Setup dof systems               |         7 |      1.97s |       1.4% |
| Setup initial conditions        |         4 |     0.457s |      0.32% |
+---------------------------------+-----------+------------+------------+
```
To me this is quite surprising: almost identical CPU and similar memory usage! I realise this is because the Kelly estimator is almost exactly balancing the CPU demands by hiking the number of cells, but that's a more dramatic change of cell numbers than I had expected.

TL;DR: this reduces CPU usage across the board, particularly in the Stokes solve, and memory by a factor of 4 (in this case). With vanilla Kelly the difference in cell numbers cancels out the extra effort and storage!

**Comment by spco** [comment]
And the pressure-pressure coupling is only in the `system_preconditioner_matrix`, as we want the weighted pressure mass matrix there.

**Comment by yinghe616** [comment]
I think that is exactly what I expected when you combine the DG and AMR together. :) Same maximum refinement level, less cells, similar amount or less of Dofs but less Dofs of Stokes Solver.

If you look at the system matrix, there still has small number seated in the pressure-pressure coupling. It looks likes to me some constraints like remove the null space? But I am not sure...

**Comment by spco** [comment]
Ah yes, you're right. TBH I don't know. Running the same example with FEM and the old logic still gives some nonzeros in the pressure-pressure block, even on 1 process. So I don't think this is a change. Perhaps somebody can give an idea - I imagine it's to do with the constraints.

**Comment by spco** [comment]
And of course the caveat to all the above is that this gives no indication of the relative accuracy being achieved!

**Comment by spco** [comment]
I've now convinced myself that the nonzeros in the pressure-pressure block are the hanging-node constraints, one per hanging node. Where a constraint is imposed, we get a nonzero on the diagonal. Simples!

**Comment by bangerth** [comment]
Yes, that's the correct explanation. The question in this context is why these entries are not there when you use DG spaces? After all, you're still using the exact same function spaces for the velocity and pressure part, right? So nothing should change there.

(The corresponding elements are generated by `make_sparsity_pattern` when passing in a constraints object.)

**Comment by spco** [comment]
I'm not sure I understand @bangerth - we do have these entries in the DG case, in the case above? 200 in the pressure-pressure block.

**Comment by bangerth** [comment]
Oh, oh, I didn't pay attention then. Yes, this looks alright then! (I moved on Monday, and we still don't have internet at the house, so dealing with github and all other things internet related has to happen in a hurry whenever I have a few minutes of connectivity...)

**Comment by bangerth** [comment]
This is ready to merge once (i) the corresponding deal.II patch is in, (ii) if you could squash everything into one commit.

Really nice improvement, by the way. Thanks for working on this!
